File: Java-OOP/Distributed_Systems/CircuitBreaker.java
/*
 * Problem: If a database goes down, 10,000 incoming requests will hang, consuming all 
 * server memory and taking the whole application down with it (Cascading Failure).
 * Logic: The Circuit Breaker. If failures cross a threshold (e.g., 3 fails), the circuit 
 * "Trips" (Opens). It immediately rejects new requests to save the server, waits 5 seconds, 
 * and then tests the connection again (Half-Open).
 */

public class CircuitBreaker {
    private enum State { CLOSED, OPEN, HALF_OPEN }
    
    private State state = State.CLOSED;
    private int failureCount = 0;
    private final int THRESHOLD = 3;
    private long lastFailureTime = 0;
    private final long RESET_TIMEOUT = 5000; // 5 seconds

    public void attemptCall(boolean serviceIsOnline) {
        if (state == State.OPEN) {
            if (System.currentTimeMillis() - lastFailureTime > RESET_TIMEOUT) {
                System.out.println("  \033[33m[Circuit HALF-OPEN]\033[0m Testing service recovery...");
                state = State.HALF_OPEN;
            } else {
                System.out.println("  \033[31m[Circuit OPEN - FAST FAIL]\033[0m Request rejected to prevent cascading failure.");
                return;
            }
        }

        if (serviceIsOnline) {
            System.out.println("  \033[32m[200 OK]\033[0m Service call successful.");
            state = State.CLOSED;
            failureCount = 0;
        } else {
            failureCount++;
            System.out.println("  \033[35m[500 ERROR]\033[0m Service call failed. (Failures: " + failureCount + "/" + THRESHOLD + ")");
            if (failureCount >= THRESHOLD) {
                System.out.println("  \033[31m[CRITICAL]\033[0m Threshold reached. TRIPPING CIRCUIT BREAKER.");
                state = State.OPEN;
                lastFailureTime = System.currentTimeMillis();
            }
        }
    }

    public static void main(String[] args) throws InterruptedException {
        System.out.println("\033[1m\n--- Distributed Circuit Breaker ---\033[0m");
        CircuitBreaker cb = new CircuitBreaker();
        
        cb.attemptCall(false);
        cb.attemptCall(false);
        cb.attemptCall(false); // Trips here
        cb.attemptCall(true);  // Fails fast
        
        System.out.println("  \033[90mWaiting 5 seconds for system cooldown...\033[0m");
        Thread.sleep(5100);
        
        cb.attemptCall(true);  // Recovers
    }
}

File: JavaScript/Resilience-Patterns/CircuitBreaker.js
/* * [PROBLEM STATEMENT]
 * If a database or third-party API goes down, continuously hammering it 
 * with requests will cause cascading system failures. We need a mechanism 
 * to "trip" and stop sending traffic to a dead service until it recovers.
 *
 * [LOGIC & INSIGHT]
 * I built an 'API Circuit Breaker'. It wraps standard functions. If a 
 * function fails 3 times in a row, the state changes from 'CLOSED' 
 * (traffic flows) to 'OPEN' (traffic is blocked). It implements a 
 * 'Cooldown Phase' where it waits before sending a test ping to see if 
 * the service is back. This is Netflix-level resilience engineering 
 * built in pure Vanilla JS.
 */

class CircuitBreaker {
    constructor(requestFunction, failureThreshold = 3, cooldownMs = 3000) {
        this.requestFunction = requestFunction;
        this.failureThreshold = failureThreshold;
        this.cooldownMs = cooldownMs;
        this.failures = 0;
        this.state = 'CLOSED'; // CLOSED = Traffic flows normally
        this.nextAttempt = null;
    }

    async fire(...args) {
        if (this.state === 'OPEN') {
            if (Date.now() > this.nextAttempt) {
                console.log(`\x1b[33m[CIRCUIT]\x1b[0m Half-Open. Testing service recovery...`);
                this.state = 'HALF_OPEN';
            } else {
                console.log(`\x1b[31m[CIRCUIT]\x1b[0m OPEN. Request blocked to prevent cascading failure.`);
                return null;
            }
        }

        try {
            const response = await this.requestFunction(...args);
            return this.onSuccess(response);
        } catch (error) {
            return this.onFailure(error);
        }
    }

    onSuccess(response) {
        this.failures = 0;
        this.state = 'CLOSED';
        console.log(`\x1b[32m[API]\x1b[0m Success:`, response);
        return response;
    }

    onFailure(error) {
        this.failures++;
        console.log(`\x1b[31m[API]\x1b[0m Request Failed. (Strike ${this.failures})`);
        
        if (this.failures >= this.failureThreshold) {
            this.state = 'OPEN';
            this.nextAttempt = Date.now() + this.cooldownMs;
            console.log(`\x1b[41m\x1b[30m [FATAL] Threshold reached. CIRCUIT TRIPPED! \x1b[0m`);
        }
        return null;
    }
}

// Mocking a failing API
const unstableAPI = async () => { throw new Error("Connection Timeout"); };

const breaker = new CircuitBreaker(unstableAPI, 2);

console.log("\x1b[1m--- INITIATING SYSTEM STRESS TEST ---\x1b[0m");
setTimeout(() => breaker.fire(), 100);
setTimeout(() => breaker.fire(), 200);
setTimeout(() => breaker.fire(), 300); // This will trip the circuit
setTimeout(() => breaker.fire(), 400); // This will be blocked instantly